Fix broken functions

raddevon opened the pull request:
These functions no longer work. When you try to create them, you run into an error: `InvalidFunctionDefinitionError: return cardinality mismatch in function declared to return exactly one value`. This fixes them by using the coalescing operator to provide fallbacks to optional properties.

I have almost 7,000 words of notes on this book that I'd like to use for a more extensive refactor at some point, but this takes care of the most pressing issue I found during my review. I'll circle back to the rest later.

vpetrovykh commented on chapter12/index.md:
```diff
@@ -48,39 +48,44 @@ Now let's `SELECT Person.strength;` and see if it works:
-    WITH opponent := assert_single((SELECT Person FILTER .name = two))
+    WITH
+        people := (SELECT Person FILTER .name IN array_unpack(people_names)),
     SELECT
-        names ++ ' win!' IF one > opponent.strength ELSE
-        two ++ ' wins!'
+        (array_join(people_names, ', ') ?? 'Fighters') ++ ' win!'
+        IF (sum(people.strength) ?? 0) > (opponent.strength ?? 0)
+        ELSE (opponent.name ?? 'Opponent') ++ ' wins!'
```
You don't need to coalesce the `sum` it already handles empty set correctly:
```
edgedb> select sum(<int64>{});
{0}
```

We also don't need to coalesce things like `array_join` because the function signature makes it required (by default), so it won't produce empty sets. However, handling the `[]` is an entirely different story and can be done with an `if`.

I think this is a cleaner version with fewer unnecessary parts:
```
WITH
    people := (SELECT Person FILTER contains(people_names, .name)),
    names := ['Weak fighters'] 
        IF people_names = <array<str>>[] 
        ELSE people_names,
SELECT
    array_join(names, ', ') ++ ' win!'
    IF sum(people.strength) > (opponent.strength ?? 0)
    ELSE (opponent.name ?? 'Opponent') ++ ' wins!'
```

vpetrovykh replied:
Actually, now that I think about it it may be unnecessary to handle `[]` of fighters because they can never win (`0 > 0` is still `false`). So it simply doesn't matter what we call them.
```
WITH
    people := (SELECT Person FILTER contains(people_names, .name)),
SELECT
    array_join(people_names, ', ') ++ ' win!'
    IF sum(people.strength) > (opponent.strength ?? 0)
    ELSE (opponent.name ?? 'Opponent') ++ ' wins!'
```

raddevon replied:
Good suggestions on the the function. I'll get those implemented.

I tried running "That's a lot of hassle, and it's not something we're ready to do right now," through the Hemingway editor which estimates readability, and it tells me that sentence is a 4th grade reading level. My feeling is that it's pretty simple, but I agree that I can go simpler than "comprehend." I'll rework that.

vpetrovykh commented on chapter12/index.md:
```diff
@@ -48,39 +48,44 @@ Now let's `SELECT Person.strength;` and see if it works:
 
+With this overload, we accept two arguments: an array of the names of the fighters and a `Person` object that is their opponent. You're seeing some new standard library functions in use here that we'll dig into more later. For now, here are the basics you need to know:
+
+- We need to call `array_unpack` when we assign the value to `people` in our `WITH`. This converts our array to a set which is needed because `IN` only works with a set.
+- At the beginning of our `SELECT`, we call `array_join` on `people_names` so that this is displayed as a string. In order to use the coalescing operator (`??`), both operands must be of the same type. Since our alternative is the string `"Fighters"`, we need this side to be a string as well.
```
These need to be re-written to account for the updated query using `contains` instead of the messier `array_unpack` and somewhat different handling of the potentially empty array of names.

vpetrovykh reviewed (changes requested):
I've made some changes to clean up the function body.

On a more general note, I think that this drifted further from the "simple Engish" format. We probably want to avoid some words like "hassle" or "comprehend". Although this is a little hard to gauge maybe.

raddevon commented:
I think I've addressed all your concerns here, @vpetrovykh. I also added a quick paragraph to explain why we provided a fallback for the single opponent's name but not for the group. Could you have another look when you get a chance and see if this ticks all the boxes?
